feat: add meta_hwm to PrivateCircuitPublicInputs

alexghr opened the pull request:
This PR adds `meta_hwm` to the context and exposes it through the private circuits so that in the last circuit the side effects created during private execution can be partitioned into reversible and irreversible effect sets

Fix #4084

just-mitch commented on yarn-project/circuits.js/src/structs/private_circuit_public_inputs.ts:
```diff
@@ -219,6 +226,7 @@ export class PrivateCircuitPublicInputs {
       fields.callContext,
       fields.argsHash,
       fields.returnValues,
+      new Fr(fields.metaHwm),
```
why does this need the Fr constructor? Why not just `fields.metaHwm`?

alexghr replied:
Oups, left over. I originally had `metaHwm` down as just a `number`. Fixed now

just-mitch commented on yarn-project/noir-protocol-circuits/src/crates/private-kernel-lib/src/private_kernel_ordering.nr:
```diff
@@ -532,4 +536,63 @@ mod tests {
+        for i in 1..5 {
+            let meta_nullifier = public_inputs.end_meta.new_nullifiers[i];
+            let nullifier = sorted_new_nullifiers[i - 1];
+            assert(meta_nullifier.eq(nullifier));
```
I'm unclear on why `sorted_new_nullifiers[0] == public_inputs.end_meta.new_nullifiers[1]` because I don't see where the `public_inputs.end_meta.new_nullifiers` are shifted by 1 to make room for the tx hash.

alexghr replied:
@just-mitch and I looked it up and discovered it's injected by the init kernel and for testing purposes, there's this block of code in the builder
https://github.com/AztecProtocol/aztec-packages/blob/bc13c0220823504ecc3162e67e3cf3b41d60b018/yarn-project/noir-protocol-circuits/src/crates/types/src/tests/previous_kernel_data_builder.nr#L47-L52

just-mitch commented on .github/workflows/protocol-circuits-gate-diff.yml:
```diff
@@ -2,9 +2,6 @@ name: Report gates diff
 
 on:
   push:
-    branches:
```
TODO: Revert this change before merge

alexghr replied:
This has been reverted

Commit:
add metaHwm to PrivateCircuitPublicInputs (TS and NR)
add meta_hwm to PrivateContext (NR)
add end_meta to KernelCircuitPublicInputsFinal (NR)
lift AggregationObject out of Combined/FinalAccumulatedData into KernelCircuitPublicInputs(Final)
update kernel_prover to partition side effects based on meta_hwm before passing into PrivateKernelOrdering circuit.

Commit:
fix serialization bug.
update snapshots.

Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>

Commit:
add meta_hwm to kernel circuit public inputs

just-mitch commented on yarn-project/noir-protocol-circuits/src/crates/types/src/abis/kernel_circuit_public_inputs.nr:
```diff
@@ -28,15 +38,22 @@ struct KernelCircuitPublicInputsBuilder {
-            end: self.end.to_final(),
+            aggregation_object: self.aggregation_object,
+            meta_hwm: self.meta_hwm,
+            end_meta: meta,
```
For clarity, this is where the extra gates are coming from. As Alex mentioned, specializing the FinalAccumulatedData to meta and app will pare back the gates significantly, and could be incorporated in this PR.

Commit:
update the snapshots

alexghr committed:
refactor: meta accumulated data (#4405)

Commit:
Merge remote-tracking branch 'origin/master' into feat/private-kernel-meta-hwm

Commit:
refactor: remove private call stack from accumulated meta data

Commit:
chore: update snapshots

spalladino commented on yarn-project/foundation/src/serialize/types.ts:
```diff
@@ -16,7 +16,7 @@ type _Tuple<T, N extends number, R extends unknown[]> = R['length'] extends N ?
 export function assertLength<T, N extends number>(array: T[], n: N): Tuple<T, N> {
   if (array.length !== n) {
-    throw new Error("Wrong 'fixed array' size");
+    throw new Error(`Wrong 'fixed array' size. Expected ${n}, got ${array.length}.`);
```
Whenever I see a change like this one I wonder how many hours of debugging were lost to lack of meaningful error messages.

spalladino reviewed (approved):
Change looks good! My only nitpick is about naming. For `AccumulatedMetaData`, I'd argue that the commitments, nullifiers, and public calls of a tx are not metadata in the strict sense, and besides, "metadata" is often too overloaded as a term that it has pretty much lost all meaning by now. I feel like "side effects" (or maybe just "effects"!) is a better name.

Same applies to `meta_hwm` or the `_PER_TX_META` constants. Maybe it's clearer if we use the term "phases" to refer to them? Without the context of this PR, it's difficult to understand from their names what they refer to.

All that said, naming is always an annoying conversation, so feel free to push this back for later.

just-mitch commented:
@spalladino I agree with with the naming nit. We picked the super generic turn `meta` since it will contain
- effects for fee prep
- enqueued calls for fee prep and fee distribution
- the transaction nullifier

We would've kept the much tighter naming tied to "fees" were it not for the inclusion of the transaction nullifier. @alexghr had advocated for something mentioning that these effects are non-revertible (or perhaps "irrevertible"). 

Very open to naming changes: the genericity muddies the water.

alexghr committed:
Merge branch 'master' into feat/private-kernel-meta-hwm

PhilWindle reviewed (commented):
Looks ok, though I'm not very familiar with many areas of the impacted code base so wouldn't be confident approving it.

When it comes to naming, I do like the idea of using 'irrevertible' to distinguish the first phase.

alexghr commented:
re naming, if it's ok we'll do it in a separate PR (this PR has already caused a bunch of merge conflicts).

I'd like to throw in another name idea into the mix: "envelope". Drawing an analogy from transactions == messages == letters, an envelope contains both a letter's metadata (or a transaction's hash) and a stamp (e.g. the fee). Thoughts?

spalladino commented:
I like it, but I'm worried that `envelope` is already a thing in Ethereum (see https://eips.ethereum.org/EIPS/eip-2718), so it may be confusing if we use it for something else.

Commit:
Merge remote-tracking branch 'origin/master' into feat/private-kernel-meta-hwm

sirasistant commented on yarn-project/noir-protocol-circuits/src/crates/types/src/constants.nr:
```diff
@@ -110,13 +114,13 @@ global CALL_CONTEXT_LENGTH: Field = 8;
-// constant as well PRIVATE_CALL_STACK_ITEM_LENGTH
-global PRIVATE_CIRCUIT_PUBLIC_INPUTS_LENGTH: Field = 204;
+// constant as well CALL_PRIVATE_FUNCTION_RETURN_SIZE
+global PRIVATE_CIRCUIT_PUBLIC_INPUTS_LENGTH: Field = 206;
```
Why 2 more? isn't just one new field?

just-mitch replied:
Thank you Alvaro! This was a mistake while resolving merge conflicts.

alexghr replied:
That's what I'm trying to figure out :laughing:. This was after a merge conflict

sirasistant reviewed (approved):
Looks good! One small question

just-mitch committed:
fix constants

just-mitch committed:
reformat solidity constants to make CI happy

just-mitch committed:
reformat solidity constants to make CI happy (take 2)

just-mitch committed:
regen snapshots and hex files. fix comment
